File: sdk/python/scrapi/actor.py
"""
Actor — main entry point for the Scrapi SDK.

Scrapi-compatible design:
  - Use as an async context manager: `async with Actor:`
  - Or call explicitly: `await Actor.init()` / `await Actor.exit()`
  - Auto-detects remote vs local mode via SCRAPI_TOKEN env var

Usage:
    import asyncio
    from scrapi import Actor, Request

    async def main():
        async with Actor:
            inp = await Actor.get_input()
            queue = await Actor.open_request_queue()
            await queue.add_request(Request.from_url(inp['url']))
            
            while not await queue.is_finished():
                req = await queue.fetch_next_request()
                if not req:
                    await asyncio.sleep(0.5)
                    continue
                await Actor.push_data({'url': req.url, 'title': '...'})
                await queue.mark_request_as_handled(req)
            
            await Actor.set_value('OUTPUT', {'status': 'done'})

    asyncio.run(main())
"""
from __future__ import annotations
import asyncio
import logging
import os
from typing import Any, Optional

from scrapi.config import config
from scrapi.models import Request
from scrapi.storages.key_value_store import KeyValueStore
from scrapi.storages.request_queue import RequestQueue
from scrapi.storages.dataset import Dataset

logger = logging.getLogger("scrapi.actor")

# ── Lazy client singleton ─────────────────────────────────────────────────────
_http_client = None

def _get_client():
    global _http_client
    if config.is_at_home and _http_client is None:
        from scrapi.client.http_client import ScrapiHttpClient
        _http_client = ScrapiHttpClient()
    return _http_client  # None in local mode


class ActorMeta(type):
    async def __aenter__(cls):
        await cls.init()
        return cls

    async def __aexit__(cls, exc_type, exc_val, exc_tb):
        await cls.exit()
        return False


class Actor(metaclass=ActorMeta):
    """
    Scrapi Actor context. Use as an async context manager or call init/exit manually.
    All methods are classmethods — use as: `await Actor.get_input()`
    """

    _initialized: bool = False
    _default_kv_store: Optional[KeyValueStore] = None
    _default_dataset: Optional[Dataset] = None
    _default_rq: Optional[RequestQueue] = None

    # ── Lifecycle ─────────────────────────────────────────────────────────

    @classmethod
    async def init(cls) -> None:
        """Initialize the actor. Called automatically by `async with Actor:`."""
        if cls._initialized:
            return
        cls._initialized = True
        mode = "remote (platform)" if config.is_at_home else "local (filesystem)"
        logger.info(f"Scrapi Actor initialized [{mode}]")

    @classmethod
    async def exit(cls, *, exit_code: int = 0) -> None:
        """Clean up resources. Called automatically by `async with Actor:`."""
        global _http_client
        if _http_client is not None:
            await _http_client.close()
            _http_client = None
        cls._initialized = False
        cls._default_kv_store = None
        cls._default_dataset = None
        cls._default_rq = None
        logger.info(f"Scrapi Actor exited (exit_code={exit_code})")

    # ── Input / Output shortcuts ──────────────────────────────────────────

    @classmethod
    async def get_input(cls) -> Optional[Any]:
        """Read INPUT key from the default Key-Value Store."""
        store = await cls._ensure_default_kv_store()
        return await store.get_value("INPUT")

    @classmethod
    async def get_value(cls, key: str) -> Optional[Any]:
        """Read any key from the default Key-Value Store."""
        store = await cls._ensure_default_kv_store()
        return await store.get_value(key)

    @classmethod
    async def set_value(
        cls, key: str, value: Any, *, content_type: Optional[str] = None
    ) -> None:
        """Write any key to the default Key-Value Store. value=None deletes the key."""
        store = await cls._ensure_default_kv_store()
        await store.set_value(key, value, content_type=content_type)

    @classmethod
    async def push_data(cls, data: Any) -> None:
        """Append rows to the default Dataset."""
        dataset = await cls._ensure_default_dataset()
        await dataset.push_data(data)

    # ── Storage openers ───────────────────────────────────────────────────

    @classmethod
    async def open_key_value_store(
        cls,
        name_or_id: Optional[str] = None,
        *,
        id: Optional[str] = None,
        name: Optional[str] = None,
        force_cloud: bool = False,
    ) -> KeyValueStore:
        """
        Open a Key-Value Store.
        
        - No args → default store for this run
        - name='my-store' → named store (create if needed)
        - id='uuid...' → open by exact ID
        - force_cloud=True → use remote even in local dev mode
        """
        client = _get_client() if (config.is_at_home or force_cloud) else None

        # Positional arg can be either name or id
        if name_or_id:
            if len(name_or_id) == 36 and "-" in name_or_id:
                id = id or name_or_id
            else:
                name = name or name_or_id

        if id:
            return KeyValueStore(id, client)

        if name:
            if client:
                result = await client.post(
                    "/api/storage/kv-stores", json={"name": name}
                )
                return KeyValueStore(result["id"], client)
            else:
                return KeyValueStore(name, client)

        # Default store
        if name_or_id is None and not id and not name:
            return await cls._ensure_default_kv_store(force_cloud=force_cloud)

        return KeyValueStore(name_or_id or "default", client)

    @classmethod
    async def open_request_queue(
        cls,
        name_or_id: Optional[str] = None,
        *,
        id: Optional[str] = None,
        name: Optional[str] = None,
        force_cloud: bool = False,
    ) -> RequestQueue:
        """Open a Request Queue."""
        client = _get_client() if (config.is_at_home or force_cloud) else None

        if name_or_id:
            if len(name_or_id) == 36 and "-" in name_or_id:
                id = id or name_or_id
            else:
                name = name or name_or_id

        if id:
            return RequestQueue(id, client)

        if name:
            if client:
                result = await client.post(
                    "/api/storage/request-queues", json={"name": name}
                )
                return RequestQueue(result["id"], client)
            else:
                return RequestQueue(name, client)

        return await cls._ensure_default_rq(force_cloud=force_cloud)

    @classmethod
    async def open_dataset(
        cls,
        name_or_id: Optional[str] = None,
        *,
        id: Optional[str] = None,
        name: Optional[str] = None,
        force_cloud: bool = False,
    ) -> Dataset:
        """Open a Dataset."""
        client = _get_client() if (config.is_at_home or force_cloud) else None

        if name_or_id:
            if len(name_or_id) == 36 and "-" in name_or_id:
                id = id or name_or_id
            else:
                name = name or name_or_id

        if id:
            return Dataset(id, client)

        if name:
            if client:
                result = await client.post("/api/storage/datasets", json={"name": name})
                return Dataset(result["id"], client)
            else:
                return Dataset(name, client)

        return await cls._ensure_default_dataset(force_cloud=force_cloud)

    # ── Internal default store management ────────────────────────────────

    @classmethod
    async def _ensure_default_kv_store(cls, force_cloud: bool = False) -> KeyValueStore:
        if cls._default_kv_store is not None:
            return cls._default_kv_store

        client = _get_client() if (config.is_at_home or force_cloud) else None

        # Use run-provided store ID if available
        store_id = config.default_kv_store_id or os.environ.get("SCRAPI_DEFAULT_KV_STORE_ID")

        if not store_id:
            if client:
                result = await client.post("/api/storage/kv-stores", json={})
                store_id = result["id"]
            else:
                store_id = "default"

        cls._default_kv_store = KeyValueStore(store_id, client)
        return cls._default_kv_store

    @classmethod
    async def _ensure_default_dataset(cls, force_cloud: bool = False) -> Dataset:
        if cls._default_dataset is not None:
            return cls._default_dataset

        client = _get_client() if (config.is_at_home or force_cloud) else None
        dataset_id = config.default_dataset_id or "default"

        if not dataset_id or dataset_id == "default":
            if client:
                result = await client.post("/api/datasets", json={})
                dataset_id = result["id"]
            else:
                dataset_id = "default"

        cls._default_dataset = Dataset(dataset_id, client)
        return cls._default_dataset

    @classmethod
    async def _ensure_default_rq(cls, force_cloud: bool = False) -> RequestQueue:
        if cls._default_rq is not None:
            return cls._default_rq

        client = _get_client() if (config.is_at_home or force_cloud) else None
        rq_id = config.default_rq_id or "default"

        if not rq_id or rq_id == "default":
            if client:
                result = await client.post("/api/storage/request-queues", json={})
                rq_id = result["id"]
            else:
                rq_id = "default"

        cls._default_rq = RequestQueue(rq_id, client)
        return cls._default_rq

File: sdk/js/src/actor.ts
/**
 * Actor — JS/TS main entry point for the Scrapi SDK.
 *
 * Usage:
 *   import { Actor } from 'scrapi-sdk';
 *   
 *   await Actor.init();
 *   const input = await Actor.getInput();
 *   const queue = await Actor.openRequestQueue();
 *   await queue.addRequest({ url: 'https://example.com', uniqueKey: 'https://example.com' });
 *   
 *   while (!await queue.isFinished()) {
 *     const req = await queue.fetchNextRequest();
 *     if (!req) { await sleep(500); continue; }
 *     await Actor.pushData({ url: req.url });
 *     await queue.markRequestAsHandled(req);
 *   }
 *   await Actor.exit();
 */
import { config } from './config';
import { ScrapiClient } from './client/scrapi_client';
import { KeyValueStore } from './storages/key_value_store';
import { RequestQueue } from './storages/request_queue';
import { Dataset } from './storages/dataset';
import { createRequest, type Request } from './models';
import { AsyncLocalStorage } from 'async_hooks';

// ── Shared Log Context (Phase 8) ───────────────────────────────────────────────
const logContext = new AsyncLocalStorage<(message: string) => void | Promise<void>>();

export class ScrapiLog {
  async info(message: string): Promise<void> {
    const cb = logContext.getStore();
    if (cb) await cb(message);
    console.log(`[INFO] ${message}`);
  }

  async error(message: string): Promise<void> {
    const cb = logContext.getStore();
    if (cb) await cb(`❌ ERROR: ${message}`);
    console.error(`[ERROR] ${message}`);
  }

  async warning(message: string): Promise<void> {
    const cb = logContext.getStore();
    if (cb) await cb(`⚠️ WARNING: ${message}`);
    console.warn(`[WARN] ${message}`);
  }

  async success(message: string): Promise<void> {
    const cb = logContext.getStore();
    if (cb) await cb(`✅ SUCCESS: ${message}`);
    console.log(`[SUCCESS] ${message}`);
  }
}


// ── Shared client singleton ────────────────────────────────────────────────────

let _client: ScrapiClient | null = null;
function getClient(forceCloud = false): ScrapiClient | null {
  if (config.isAtHome || forceCloud) {
    if (!_client) _client = new ScrapiClient();
    return _client;
  }
  return null;
}

// ── Actor class ───────────────────────────────────────────────────────────────

export class Actor {
  private static _initialized = false;
  private static _defaultKvStore: KeyValueStore | null = null;
  private static _defaultDataset: Dataset | null = null;
  private static _defaultRq: RequestQueue | null = null;

  /**
   * SDK Logger (Phase 8)
   */
  public static log = new ScrapiLog();


  // ── Lifecycle ─────────────────────────────────────────────────────────────

  static async init(): Promise<void> {
    if (Actor._initialized) return;
    Actor._initialized = true;
    const mode = config.isAtHome ? 'remote (platform)' : 'local (filesystem)';
    console.log(`[Scrapi] Actor initialized [${mode}]`);
  }

  static async exit(options?: { exitCode?: number }): Promise<void> {
    _client = null;
    Actor._initialized = false;
    Actor._defaultKvStore = null;
    Actor._defaultDataset = null;
    Actor._defaultRq = null;
    console.log(`[Scrapi] Actor exited (exitCode=${options?.exitCode ?? 0})`);
  }

  /**
   * Internal use: set the callback for real-time log streaming.
   * Links Actor.log calls to the platform's run stream.
   */
  static setLogCallback(callback: (message: string) => void | Promise<void>, fn: () => Promise<void>): Promise<void> {
    return logContext.run(callback, fn);
  }


  // ── Input / Output shortcuts ───────────────────────────────────────────────

  static async getInput<T = Record<string, unknown>>(): Promise<T | null> {
    const store = await Actor._ensureDefaultKvStore();
    return store.getValue<T>('INPUT');
  }

  static async getValue<T = unknown>(key: string): Promise<T | null> {
    const store = await Actor._ensureDefaultKvStore();
    return store.getValue<T>(key);
  }

  static async setValue(key: string, value: unknown, options?: { contentType?: string }): Promise<void> {
    const store = await Actor._ensureDefaultKvStore();
    await store.setValue(key, value, options);
  }

  static async pushData(data: Record<string, unknown> | Record<string, unknown>[]): Promise<void> {
    const dataset = await Actor._ensureDefaultDataset();
    await dataset.pushData(Array.isArray(data) ? data : [data]);
  }

  // ── Storage openers ───────────────────────────────────────────────────────

  static async openKeyValueStore(
    nameOrId?: string,
    options?: { forceCloud?: boolean }
  ): Promise<KeyValueStore> {
    const client = getClient(options?.forceCloud);

    if (!nameOrId) return Actor._ensureDefaultKvStore(options?.forceCloud);

    // Looks like a UUID → treat as ID
    if (/^[0-9a-f-]{36}$/i.test(nameOrId)) return new KeyValueStore(nameOrId, client);

    // Named store
    if (client) {
      const result = await client.post<any>('/api/storage/kv-stores', { name: nameOrId });
      return new KeyValueStore(result.id, client);
    }
    return new KeyValueStore(nameOrId, null);
  }

  static async openRequestQueue(
    nameOrId?: string,
    options?: { forceCloud?: boolean }
  ): Promise<RequestQueue> {
    const client = getClient(options?.forceCloud);

    if (!nameOrId) return Actor._ensureDefaultRq(options?.forceCloud);

    if (/^[0-9a-f-]{36}$/i.test(nameOrId)) return new RequestQueue(nameOrId, client);

    if (client) {
      const result = await client.post<any>('/api/storage/request-queues', { name: nameOrId });
      return new RequestQueue(result.id, client);
    }
    return new RequestQueue(nameOrId, null);
  }

  static async openDataset<T extends Record<string, unknown> = Record<string, unknown>>(
    nameOrId?: string,
    options?: { forceCloud?: boolean }
  ): Promise<Dataset<T>> {
    const client = getClient(options?.forceCloud);

    if (!nameOrId) return Actor._ensureDefaultDataset(options?.forceCloud) as Promise<Dataset<T>>;

    if (/^[0-9a-f-]{36}$/i.test(nameOrId)) return new Dataset<T>(nameOrId, client);

    if (client) {
      const result = await client.post<any>('/api/storage/datasets', { name: nameOrId });
      return new Dataset<T>(result.id, client);
    }
    return new Dataset<T>(nameOrId, null);
  }

  // ── Internal default store management ─────────────────────────────────────

  private static async _ensureDefaultKvStore(forceCloud = false): Promise<KeyValueStore> {
    if (Actor._defaultKvStore) return Actor._defaultKvStore;
    const client = getClient(forceCloud);
    let storeId = config.defaultKvStoreId;
    if (!storeId) {
      if (client) {
        const r = await client.post<any>('/api/storage/kv-stores', {});
        storeId = r.id;
      } else {
        storeId = 'default';
      }
    }
    Actor._defaultKvStore = new KeyValueStore(storeId!, client);
    return Actor._defaultKvStore;
  }

  private static async _ensureDefaultDataset(forceCloud = false): Promise<Dataset> {
    if (Actor._defaultDataset) return Actor._defaultDataset;
    const client = getClient(forceCloud);
    let id = config.defaultDatasetId;
    if (!id) {
      if (client) {
        const r = await client.post<any>('/api/storage/datasets', {});
        id = r.id;
      } else {
        id = 'default';
      }
    }
    Actor._defaultDataset = new Dataset(id!, client);
    return Actor._defaultDataset;
  }

  private static async _ensureDefaultRq(forceCloud = false): Promise<RequestQueue> {
    if (Actor._defaultRq) return Actor._defaultRq;
    const client = getClient(forceCloud);
    let id = config.defaultRqId;
    if (!id) {
      if (client) {
        const r = await client.post<any>('/api/storage/request-queues', {});
        id = r.id;
      } else {
        id = 'default';
      }
    }
    Actor._defaultRq = new RequestQueue(id!, client);
    return Actor._defaultRq;
  }
}
